torch/csrc/jit/passes/symbolic_shape_analysis.cpp

```diff
@@ -205,206 +268,90 @@ void replaceWithIValue(Value* v, IValue val) {
+      graph_in_var->setType(val_type);
     }
-
-    // we can't optimize a tensor without fixed rank
-    if (symbolic_shapes.rank()) {
-      node_symbolic_input_indices_.emplace_back(
-          shape_compute_graph_index, symbolic_shapes);
+    // We handle non-constant values in the shape propagation step
+    if (IValue* cur_val = c10::get_if<IValue>(&argument)) {
+      GRAPH_DEBUG("Substituting constant input ", *cur_val);
+      replaceWithIValue(graph_in_var, *cur_val);
     }
   }
 
-  // returns partially evaluated shape compute graph
-  std::shared_ptr<Graph> run() {
-    bool made_change = true;
-    constexpr size_t MAX_ATTEMPTS = 8;
-    size_t curr_attempt = 0;
-    while (made_change && curr_attempt < MAX_ATTEMPTS) {
-      curr_attempt++;
-      // symbolic shape concrete values are only used in final shape extraction
-      substituteInputTensorProperties(/*symbolic_shape_values*/ nullptr);
-      made_change = shapeGraphCleanupPasses(shape_compute_graph_);
+  void substituteConstantInputs() {
+    if (schema_->name() == "aten::cat") {
+      // Modifying the graph where _node is part of to not use the tensor
+      // construct
+
+      // When we have partially evaluate a list of Tensors like cat(tensor[])
+      // We have a few problems:
+      // - optimizing out calls to the length of the list: len(tensors)
+      // - resolving accesses of the list to the tensor symbolic sizes the
+      // corresponding list element We can solve both of these problems by
+      // replacing the partial evaluation of cat([x, y]) def cat(tensors:
+      // List[List[int]], dim: int)
+      //    body
+      // with
+      // def cat(x, y, dim: int)
+      //     tensors = [x, y]
+      //     body
+      uint64_t li_length = inputs_.size() - (schema_->arguments().size() - 1);
+      std::vector<Value*> li_inputs;
+      Value* graph_input = shape_compute_graph_->inputs().at(0);
+      for (size_t j = 0; j < li_length; ++j) {
+        auto new_inp = shape_compute_graph_->insertInput(j);
+        new_inp->setType(ListType::ofInts());
+        li_inputs.push_back(new_inp);
+      }
+      WithInsertPoint guard(*shape_compute_graph_->block()->nodes().begin());
+      auto new_li = shape_compute_graph_->insertNode(
+          shape_compute_graph_->createList(ListType::ofInts(), li_inputs));
+      graph_input->replaceAllUsesWith(new_li->output());
+
+      shape_compute_graph_->eraseInput(li_length);
     }
-    std::unordered_map<Value*, int64_t> symbolic_shape_values;
-    substituteInputTensorProperties(&symbolic_shape_values);
-    GRAPH_DUMP("Done with partial evaluation", shape_compute_graph_);
 
-    extractOutputShape(symbolic_shape_values);
-    return shape_compute_graph_;
+    for (size_t op_in_index = 0;
+         op_in_index < shape_compute_graph_->inputs().size();
+         op_in_index++) {
+      substituteInput(
+          inputs_[op_in_index], shape_compute_graph_->inputs().at(op_in_index));
```

Comment:
This is changing behavior - we used to be e.g. refining an Optional Input into the shape compute graph to a non-optional if the Value* that was being used wasn't optional. Now we are just looking at the types of the shape compute graph which won't do anything